test/e2e/upgrade_cluster_test.go

```diff
@@ -0,0 +1,189 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/onsi/gomega"
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	nodePoolAnnotationCurrentConfig        = "hypershift.openshift.io/nodePoolCurrentConfig"
+	nodePoolAnnotationCurrentConfigVersion = "hypershift.openshift.io/nodePoolCurrentConfigVersion"
+	hostedClusterUpgradeTestLabel          = "hypershift.openshift.io/upgrade-test"
+)
+
+// TestUpgradeHyperShift validates that a HyperShift Operator upgrade won't unnecessarily cause a node rollout.
+// The test must be triggered via the flag upgrade.run-tests and the HyperShift Operator initial image must be set
+// via the flag e2e.hypershift-operator-initial-image.
+func TestUpgradeHyperShift(t *testing.T) {
+	if globalOpts.RunUpgradeTest {
+		var once sync.Once
```

Comment:
nit: flip if/else so its easier to read.
```suggestion
func TestUpgradeHyperShift(t *testing.T) {
	if !globalOpts.RunUpgradeTest {
		// This test should be triggered intentionally by setting the option above,
		// otherwise skip it.
		t.SkipNow()
           }
           // test code
           .....
}
```

Reply:
Addressed